<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>7 Outputs | BiocMAP</title>
  <meta name="description" content="7 Outputs | BiocMAP" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="7 Outputs | BiocMAP" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 Outputs | BiocMAP" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="inputs.html"/>
<link rel="next" href="configuration.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Overview</a>
<ul>
<li><a href="index.html#cite-biocmap">Cite <code>BiocMAP</code></a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i>R session information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>1</b> Quick Start</a>
<ul>
<li class="chapter" data-level="1.1" data-path="quick-start.html"><a href="quick-start.html#setup"><i class="fa fa-check"></i><b>1.1</b> Setup</a></li>
<li class="chapter" data-level="1.2" data-path="quick-start.html"><a href="quick-start.html#quick-config"><i class="fa fa-check"></i><b>1.2</b> Configuration</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="quick-start.html"><a href="quick-start.html#your-main-script"><i class="fa fa-check"></i><b>1.2.1</b> Your “main” script</a></li>
<li class="chapter" data-level="1.2.2" data-path="quick-start.html"><a href="quick-start.html#your-config-file"><i class="fa fa-check"></i><b>1.2.2</b> Your config file</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="setup-details.html"><a href="setup-details.html"><i class="fa fa-check"></i><b>2</b> Setup Details</a>
<ul>
<li class="chapter" data-level="2.1" data-path="setup-details.html"><a href="setup-details.html#requirements"><i class="fa fa-check"></i><b>2.1</b> Requirements</a></li>
<li class="chapter" data-level="2.2" data-path="setup-details.html"><a href="setup-details.html#installation"><i class="fa fa-check"></i><b>2.2</b> Installation</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="setup-details.html"><a href="setup-details.html#troubleshooting"><i class="fa fa-check"></i><b>2.2.1</b> Troubleshooting</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="setup-details.html"><a href="setup-details.html#run-the-pipeline"><i class="fa fa-check"></i><b>2.3</b> Run the Pipeline</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="setup-details.html"><a href="setup-details.html#run-in-a-slurm-environment-cluster"><i class="fa fa-check"></i><b>2.3.1</b> Run in a SLURM environment/ cluster</a></li>
<li class="chapter" data-level="2.3.2" data-path="setup-details.html"><a href="setup-details.html#run-on-a-sun-grid-engines-sge-cluster"><i class="fa fa-check"></i><b>2.3.2</b> Run on a Sun Grid Engines (SGE) cluster</a></li>
<li class="chapter" data-level="2.3.3" data-path="setup-details.html"><a href="setup-details.html#run-locally"><i class="fa fa-check"></i><b>2.3.3</b> Run locally</a></li>
<li class="chapter" data-level="2.3.4" data-path="setup-details.html"><a href="setup-details.html#run-on-the-jhpce-cluster"><i class="fa fa-check"></i><b>2.3.4</b> Run on the <span>JHPCE</span> cluster</a></li>
<li class="chapter" data-level="2.3.5" data-path="setup-details.html"><a href="setup-details.html#example-main-script"><i class="fa fa-check"></i><b>2.3.5</b> Example main script</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="setup-details.html"><a href="setup-details.html#sharing"><i class="fa fa-check"></i><b>2.4</b> Sharing the pipeline with many users</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="setup-details.html"><a href="setup-details.html#customizing-execution-for-each-user"><i class="fa fa-check"></i><b>2.4.1</b> Customizing execution for each user</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="command-opts.html"><a href="command-opts.html"><i class="fa fa-check"></i><b>3</b> All Command Options</a>
<ul>
<li class="chapter" data-level="3.1" data-path="command-opts.html"><a href="command-opts.html#biocmap-options"><i class="fa fa-check"></i><b>3.1</b> BiocMAP Options</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="command-opts.html"><a href="command-opts.html#mandatory-parameters"><i class="fa fa-check"></i><b>3.1.1</b> Mandatory Parameters</a></li>
<li class="chapter" data-level="3.1.2" data-path="command-opts.html"><a href="command-opts.html#optional-parameters"><i class="fa fa-check"></i><b>3.1.2</b> Optional Parameters</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="command-opts.html"><a href="command-opts.html#next-opts"><i class="fa fa-check"></i><b>3.2</b> Nextflow Options</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="pipeline-overview.html"><a href="pipeline-overview.html"><i class="fa fa-check"></i><b>4</b> Pipeline Overview</a>
<ul>
<li class="chapter" data-level="4.1" data-path="pipeline-overview.html"><a href="pipeline-overview.html#preparation-steps"><i class="fa fa-check"></i><b>4.1</b> Preparation Steps</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="pipeline-overview.html"><a href="pipeline-overview.html#downloading-and-preparing-required-annotation-files"><i class="fa fa-check"></i><b>4.1.1</b> Downloading and preparing required annotation files</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="pipeline-overview.html"><a href="pipeline-overview.html#main-workflow-steps-first-module"><i class="fa fa-check"></i><b>4.2</b> Main Workflow Steps: First Module</a></li>
<li class="chapter" data-level="4.3" data-path="pipeline-overview.html"><a href="pipeline-overview.html#main-workflow-steps-second-module"><i class="fa fa-check"></i><b>4.3</b> Main Workflow Steps: Second Module</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="annotation.html"><a href="annotation.html"><i class="fa fa-check"></i><b>5</b> Annotation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="annotation.html"><a href="annotation.html#default-annotation"><i class="fa fa-check"></i><b>5.1</b> Default Annotation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="annotation.html"><a href="annotation.html#choosing-a-release"><i class="fa fa-check"></i><b>5.1.1</b> Choosing a release</a></li>
<li class="chapter" data-level="5.1.2" data-path="annotation.html"><a href="annotation.html#choosing-build"><i class="fa fa-check"></i><b>5.1.2</b> Choosing a “build”</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="annotation.html"><a href="annotation.html#custom-annotation"><i class="fa fa-check"></i><b>5.2</b> Custom Annotation</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="inputs.html"><a href="inputs.html"><i class="fa fa-check"></i><b>6</b> Inputs</a>
<ul>
<li class="chapter" data-level="6.1" data-path="inputs.html"><a href="inputs.html#the-samples.manifest-file"><i class="fa fa-check"></i><b>6.1</b> The <code>samples.manifest</code> File</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="inputs.html"><a href="inputs.html#what-samples.manifest-should-look-like"><i class="fa fa-check"></i><b>6.1.1</b> What <code>samples.manifest</code> should look like</a></li>
<li class="chapter" data-level="6.1.2" data-path="inputs.html"><a href="inputs.html#more-details-regarding-inputs"><i class="fa fa-check"></i><b>6.1.2</b> More details regarding inputs</a></li>
<li class="chapter" data-level="6.1.3" data-path="inputs.html"><a href="inputs.html#creating-a-manifest-file"><i class="fa fa-check"></i><b>6.1.3</b> Creating a manifest file</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="inputs.html"><a href="inputs.html#the-rules.txt-file"><i class="fa fa-check"></i><b>6.2</b> The <code>rules.txt</code> File</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="inputs.html"><a href="inputs.html#what-rules.txt-should-look-like"><i class="fa fa-check"></i><b>6.2.1</b> What <code>rules.txt</code> should look like</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="outputs.html"><a href="outputs.html"><i class="fa fa-check"></i><b>7</b> Outputs</a>
<ul>
<li class="chapter" data-level="7.1" data-path="outputs.html"><a href="outputs.html#main-outputs"><i class="fa fa-check"></i><b>7.1</b> Main Outputs</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="outputs.html"><a href="outputs.html#cpg-bsseq-object"><i class="fa fa-check"></i><b>7.1.1</b> CpG <em>bsseq</em> object</a></li>
<li class="chapter" data-level="7.1.2" data-path="outputs.html"><a href="outputs.html#cph-bsseq-object"><i class="fa fa-check"></i><b>7.1.2</b> CpH <em>bsseq</em> object</a></li>
<li class="chapter" data-level="7.1.3" data-path="outputs.html"><a href="outputs.html#storage-method"><i class="fa fa-check"></i><b>7.1.3</b> Storage method</a></li>
<li class="chapter" data-level="7.1.4" data-path="outputs.html"><a href="outputs.html#metrics"><i class="fa fa-check"></i><b>7.1.4</b> Metrics</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="outputs.html"><a href="outputs.html#intermediary-outputs"><i class="fa fa-check"></i><b>7.2</b> Intermediary Outputs</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="outputs.html"><a href="outputs.html#first-module"><i class="fa fa-check"></i><b>7.2.1</b> First Module</a></li>
<li class="chapter" data-level="7.2.2" data-path="outputs.html"><a href="outputs.html#second-module"><i class="fa fa-check"></i><b>7.2.2</b> Second Module</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="configuration.html"><a href="configuration.html"><i class="fa fa-check"></i><b>8</b> Configuration</a>
<ul>
<li class="chapter" data-level="8.1" data-path="configuration.html"><a href="configuration.html#specifying-options-for-your-cluster"><i class="fa fa-check"></i><b>8.1</b> Specifying Options for your Cluster</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="configuration.html"><a href="configuration.html#time"><i class="fa fa-check"></i><b>8.1.1</b> Time</a></li>
<li class="chapter" data-level="8.1.2" data-path="configuration.html"><a href="configuration.html#cluster-specific-options"><i class="fa fa-check"></i><b>8.1.2</b> Cluster-specific options</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="configuration.html"><a href="configuration.html#biocmap-settings"><i class="fa fa-check"></i><b>8.2</b> BiocMAP settings</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="configuration.html"><a href="configuration.html#annotation-settings"><i class="fa fa-check"></i><b>8.2.1</b> Annotation settings</a></li>
<li class="chapter" data-level="8.2.2" data-path="configuration.html"><a href="configuration.html#arioc-settings"><i class="fa fa-check"></i><b>8.2.2</b> Arioc settings</a></li>
<li class="chapter" data-level="8.2.3" data-path="configuration.html"><a href="configuration.html#miscellaneous-settings"><i class="fa fa-check"></i><b>8.2.3</b> Miscellaneous settings</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="software.html"><a href="software.html"><i class="fa fa-check"></i><b>9</b> Software</a>
<ul>
<li class="chapter" data-level="9.1" data-path="software.html"><a href="software.html#software-versions"><i class="fa fa-check"></i><b>9.1</b> Software Versions</a></li>
<li class="chapter" data-level="9.2" data-path="software.html"><a href="software.html#docker-images"><i class="fa fa-check"></i><b>9.2</b> Docker Images</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="help.html"><a href="help.html"><i class="fa fa-check"></i><b>10</b> Help</a>
<ul>
<li class="chapter" data-level="10.1" data-path="help.html"><a href="help.html#common-errors"><i class="fa fa-check"></i><b>10.1</b> Common Errors</a></li>
<li class="chapter" data-level="10.2" data-path="help.html"><a href="help.html#resuming"><i class="fa fa-check"></i><b>10.2</b> Resuming</a></li>
<li class="chapter" data-level="10.3" data-path="help.html"><a href="help.html#docker-help"><i class="fa fa-check"></i><b>10.3</b> Docker help</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="example-analysis.html"><a href="example-analysis.html"><i class="fa fa-check"></i><b>11</b> Example Analysis</a>
<ul>
<li class="chapter" data-level="11.1" data-path="example-analysis.html"><a href="example-analysis.html#add-experiment-metadata-to-biocmap-outputs"><i class="fa fa-check"></i><b>11.1</b> Add Experiment Metadata to BiocMAP Outputs</a></li>
<li class="chapter" data-level="11.2" data-path="example-analysis.html"><a href="example-analysis.html#exploratory-plots"><i class="fa fa-check"></i><b>11.2</b> Exploratory Plots</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="example-analysis.html"><a href="example-analysis.html#bisulfite-conversion-efficiency-by-cell-population"><i class="fa fa-check"></i><b>11.2.1</b> Bisulfite-Conversion Efficiency by Cell Population</a></li>
<li class="chapter" data-level="11.2.2" data-path="example-analysis.html"><a href="example-analysis.html#relationship-between-methylation-fractions-across-cytosine-context-by-cell-population"><i class="fa fa-check"></i><b>11.2.2</b> Relationship between Methylation Fractions across Cytosine Context, by Cell Population</a></li>
<li class="chapter" data-level="11.2.3" data-path="example-analysis.html"><a href="example-analysis.html#fraction-of-covered-cytosines-by-cell-population"><i class="fa fa-check"></i><b>11.2.3</b> Fraction of Covered Cytosines by Cell Population</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BiocMAP</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="outputs" class="section level1" number="7">
<h1><span class="header-section-number">7</span> Outputs</h1>
<div id="main-outputs" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> Main Outputs</h2>
<div class="figure">
<img src="images/bsseq_output_objects.png" alt="" />
<p class="caption">Main Outputs</p>
</div>
<p>The major outputs from the second module are R objects from the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq.html"><em>bsseq</em> Bioconductor package</a>, which contain methylation proportion and coverage information at all cytosine loci in the human genome. bsseq extends the <code>SummarizedExperiment</code> class, which provides a general and popular format for storing genomics data. Two <code>bsseq</code> objects are produced, with one object containing cytosine sites in CpG context, and the other containing the remaining CpH loci.</p>
<div id="cpg-bsseq-object" class="section level3" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> CpG <em>bsseq</em> object</h3>
<p>We “strand collapse” CpG loci, which involves combining methylation data from both genomic strands (and thus discarding strand-specific information). The object is “smoothed” with the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r83">BSmooth algorithm</a>, a process for inferring stable regional estimates of methylation levels. Loci are ordered by genomic position.</p>
</div>
<div id="cph-bsseq-object" class="section level3" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> CpH <em>bsseq</em> object</h3>
<p>We retain strand-specific information for CpH loci. These loci are also ordered by genomic position.</p>
</div>
<div id="storage-method" class="section level3" number="7.1.3">
<h3><span class="header-section-number">7.1.3</span> Storage method</h3>
<p>Because all cytosines in the genome are included in the output objects, the data may occupy tens or even hundreds of gigabytes in memory (RAM) if loaded in a typical fashion. To enable working with the objects in a reasonable amount of memory, the assays (in this case methylation fraction and coverage counts) are HDF5-backed using the <a href="https://bioconductor.org/packages/release/bioc/html/HDF5Array.html">HDF5Array R package</a>, based on the <a href="https://bioconductor.org/packages/release/bioc/html/DelayedArray.html">DelayedArray</a> framework. Essentially, this involves storing the assays in a .h5 file, a format designed to enable working with on-disk data as if it were loaded in RAM.</p>
</div>
<div id="metrics" class="section level3" number="7.1.4">
<h3><span class="header-section-number">7.1.4</span> Metrics</h3>
<p>The <code>colData</code> slot of each <code>bsseq</code> object includes metrics collected from various processing steps throughout the pipeline. These metrics are also included in a standalone RDA file containing an R data frame whose columns are different quality metrics, and whose rows are associated with sample names. A list of the exact column names and their descriptions is given below.</p>
<table>
<colgroup>
<col width="25%" />
<col width="25%" />
<col width="50%" />
</colgroup>
<thead>
<tr class="header">
<th>Metric Name</th>
<th>Processing step</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>FQCbasicStats</td>
<td>FastQC</td>
<td>Value for “Basic Statistics” in FastQC summary output</td>
</tr>
<tr class="even">
<td>perBaseQual</td>
<td>FastQC</td>
<td>Value for “Per base sequence quality” in FastQC summary output</td>
</tr>
<tr class="odd">
<td>perTileQual</td>
<td>FastQC</td>
<td>Value for “Per tile sequence quality” in FastQC summary output</td>
</tr>
<tr class="even">
<td>perSeqQual</td>
<td>FastQC</td>
<td>Value for “Per sequence quality scores” in FastQC summary output</td>
</tr>
<tr class="odd">
<td>perBaseContent</td>
<td>FastQC</td>
<td>Value for “Per base sequence content” in FastQC summary output</td>
</tr>
<tr class="even">
<td>GCcontent</td>
<td>FastQC</td>
<td>Value for “Per sequence GC content” in FastQC summary output</td>
</tr>
<tr class="odd">
<td>Ncontent</td>
<td>FastQC</td>
<td>Value for “Per base N content” in FastQC summary output</td>
</tr>
<tr class="even">
<td>SeqLengthDist</td>
<td>FastQC</td>
<td>Value for “Sequence Length Distribution” in FastQC summary output</td>
</tr>
<tr class="odd">
<td>SeqDuplication</td>
<td>FastQC</td>
<td>Value for “Sequence Duplication Levels” in FastQC summary output</td>
</tr>
<tr class="even">
<td>OverrepSeqs</td>
<td>FastQC</td>
<td>Value for “Overrepresented sequences” in FastQC summary output</td>
</tr>
<tr class="odd">
<td>AdapterContent</td>
<td>FastQC</td>
<td>Value for “Adapter Content” in FastQC summary output</td>
</tr>
<tr class="even">
<td>KmerContent</td>
<td>FastQC</td>
<td>Value for “Kmer Content” in FastQC summary output</td>
</tr>
<tr class="odd">
<td>Total_reads_processed</td>
<td>Trim Galore!</td>
<td>Value for “Total reads processed” entry in STDOUT from trimming</td>
</tr>
<tr class="even">
<td>Reads_with_adapters</td>
<td>Trim Galore!</td>
<td>Value for “Reads with adapters” entry in STDOUT from trimming</td>
</tr>
<tr class="odd">
<td>Reads_written_passing_filters</td>
<td>Trim Galore!</td>
<td>Value for “Reads written (passing filters)” entry in STDOUT from trimming</td>
</tr>
<tr class="even">
<td>Total_basepairs_processed</td>
<td>Trim Galore!</td>
<td>Value for “Total basepairs processed” entry in STDOUT from trimming</td>
</tr>
<tr class="odd">
<td>Quality_trimmed</td>
<td>Trim Galore!</td>
<td>Value for “Quality-trimmed” entry in STDOUT from trimming</td>
</tr>
<tr class="even">
<td>Total_written_filtered</td>
<td>Trim Galore!</td>
<td>Value for “Total written (filtered)” entry in STDOUT from trimming</td>
</tr>
<tr class="odd">
<td>Sequence</td>
<td>Trim Galore!</td>
<td>Detected adapter sequence when trimming</td>
</tr>
<tr class="even">
<td>was_trimmed</td>
<td>Trim Galore!</td>
<td>Logical value (“TRUE” or “FALSE”) indicating whether this sample was trimmed</td>
</tr>
<tr class="odd">
<td>pairs</td>
<td>Arioc</td>
<td>Total number of pairs present in input data</td>
</tr>
<tr class="even">
<td>conc_pairs_total</td>
<td>Arioc</td>
<td>Total number of concordantly aligned pairs</td>
</tr>
<tr class="odd">
<td>conc_pairs_1_mapping</td>
<td>Arioc</td>
<td>Number of uniquely mapped concordant alignments</td>
</tr>
<tr class="even">
<td>conc_pairs_many_mappings</td>
<td>Arioc</td>
<td>Number of non-uniquely mapped concordant pairs</td>
</tr>
<tr class="odd">
<td>disc_pairs</td>
<td>Arioc</td>
<td>Number of pairs that aligned discordantly</td>
</tr>
<tr class="even">
<td>rejected_pairs</td>
<td>Arioc</td>
<td>Number of pairs that were “rejected” during alignment</td>
</tr>
<tr class="odd">
<td>unmapped_pairs</td>
<td>Arioc</td>
<td>Number of pairs that didn’t both successfully map</td>
</tr>
<tr class="even">
<td>mates_not_in_paired_maps_total</td>
<td>Arioc</td>
<td>Value for “mates not in paired mappings” field in “SAM output” section of verbose output from alignment</td>
</tr>
<tr class="odd">
<td>mates_NIPM_with_no_maps</td>
<td>Arioc</td>
<td>Number of mates not in paired mappings having no mappings</td>
</tr>
<tr class="even">
<td>mates_NIPM_with_1_map</td>
<td>Arioc</td>
<td>Number of mates not in paired mappings having 1 mapping</td>
</tr>
<tr class="odd">
<td>mates_NIPM_with_many_maps</td>
<td>Arioc</td>
<td>Number of mates not in paired mappings having 2 or more mappings</td>
</tr>
<tr class="even">
<td>total_mapped_mates</td>
<td>Arioc</td>
<td>Total number of mates mapped</td>
</tr>
<tr class="odd">
<td>duplicate_maps</td>
<td>Arioc</td>
<td>Value for “duplicate mappings (unreported)” field in “SAM output” section of verbose output from alignment</td>
</tr>
<tr class="even">
<td>maxQlen</td>
<td>Arioc</td>
<td>Value for “maximum Q length” field in “SAM output” section of verbose output from alignment</td>
</tr>
<tr class="odd">
<td>max_diag_band_width</td>
<td>Arioc</td>
<td>Value for “maximum diagonal band width” field in “SAM output” section of verbose output from alignment</td>
</tr>
<tr class="even">
<td>TLEN mean</td>
<td>Arioc</td>
<td>Value for “TLEN mean (stdev, skewness)” field in “SAM output” section of verbose output from alignment</td>
</tr>
<tr class="odd">
<td>perc_M_CpG</td>
<td>Methylation extraction</td>
<td>Among all cytosines observed in CpG context, the percent that are methylated</td>
</tr>
<tr class="even">
<td>perc_M_CHG</td>
<td>Methylation extraction</td>
<td>Among all cytosines observed in CHG context, the percent that are methylated</td>
</tr>
<tr class="odd">
<td>perc_M_CHH</td>
<td>Methylation extraction</td>
<td>Among all cytosines observed in CHH context, the percent that are methylated</td>
</tr>
<tr class="even">
<td>lambda_bs_conv_eff</td>
<td>Lambda pseudo-alignment</td>
<td>Estimated bisulfite conversion efficiency as a percentage</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="intermediary-outputs" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Intermediary Outputs</h2>
<p>BiocMAP generates a number of files along the process before producing the main outputs of interest in each module. Each of these “intermediary” files is described below.</p>
<div id="first-module" class="section level3" number="7.2.1">
<h3><span class="header-section-number">7.2.1</span> First Module</h3>
<p><strong>Preprocessing Logs</strong> <code>preprocessing/</code></p>
<ul>
<li><code>preprocess_input_first_half.log</code>: Information about how the input <code>samples.manifest</code> file was parsed to internally handle input files correctly downstream.</li>
</ul>
<p><strong>FastQC Outputs</strong> <code>fastQC/</code></p>
<ul>
<li><code>[trim_status]/[file_name]/*</code>: Outputs from <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>. Here <code>[trim_status]</code> indicates when FastQC was performed: <code>Untrimmed</code> is before trimming, and <code>Trimmed</code> is after. <code>[file_name]</code> contains the sample name, and if applicable, the mate number.</li>
</ul>
<p><strong>Trimmed FASTQ Files</strong> <code>trimming/</code></p>
<ul>
<li><code>[sample_name]_trimmed*.fastq</code>: Trimmed FASTQ files, if applicable, from <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim Galore</a>.</li>
</ul>
<p><strong>Raw Alignment-related Files</strong> <code>Arioc/</code></p>
<ul>
<li><code>sams/[sample_name].c.sam</code>: Concordant alignments output directly from Arioc, in SAM format.</li>
<li><code>configs/[sample_name]_align.cfg</code>: The exact config file used by Arioc to align reads to the reference genome.</li>
<li><code>configs/[sample_name]_encode.cfg</code>: The exact config file used by Arioc to encode reads prior to alignment to the reference genome.</li>
<li><code>logs/*_[sample_name].log</code>: Logs from encoding and aligning samples with Arioc.</li>
</ul>
<p><strong>Filtered Alignments</strong> <code>FilteredAlignments/bams</code></p>
<ul>
<li><code>[sample_name].cfu.sorted.bam</code> and <code>[sample_name].cfu.sorted.bam.bai</code>: Coordinate-sorted, quality-filtered (only including alignments with <code>MAPQ</code> &gt;= 5), unique (duplicate mappings removed) alignments in BAM format, with a corresponding BAM index.</li>
</ul>
</div>
<div id="second-module" class="section level3" number="7.2.2">
<h3><span class="header-section-number">7.2.2</span> Second Module</h3>
<p><strong>Preprocessing Logs</strong> <code>preprocessing/</code></p>
<ul>
<li><code>preprocess_input_second_half.log</code>: Information about how the input <code>rules.txt</code> file was parsed to internally handle input files correctly downstream.</li>
</ul>
<p><strong>Lambda-based Bisulfite-Conversion Estimatation</strong> <code>lambda/</code></p>
<ul>
<li><code>[sample_name]_lambda_pseudo.log</code>: (For experiments with lambda spike-ins, using the <code>--with_lambda</code> <a href="command-opts.html#command-opts">option</a>) Logs containing a percentage estimate for the bisulfite-conversion efficiency for each sample, based on pseudo-aligning input FASTQs to both the lambda bacteriophage transcriptome and an “<em>in silico</em> bisulfite-converted copy” of the transcriptome, and comparing alignment rates.</li>
</ul>
<p><strong>Bismark Methylation Extractor Outputs</strong> <code>BME/</code></p>
<ul>
<li><code>[sample_name]/*</code>: (When using the <code>--use_bme</code> <a href="command-opts.html#command-opts">option</a>) Files output from running <code>bismark_methylation_extractor</code>, including compressed text-based methylation information, a splitting report, and an “M-bias” report.</li>
</ul>
<p><strong>Cytosine Reports and Bedgraphs</strong> <code>Reports/</code></p>
<ul>
<li><p><code>[sample_name]/[sample_name].*.CX_report.txt</code>: “Cytosine reports” containing methylation counts and context for every cytosine in the genome, split by chromosome. By default, these are generated with <code>MethylDackel</code>, or otherwise with <code>coverage2cytosine</code> from the <code>Bismark</code> software suite when using the <code>--use_bme</code> <a href="command-opts.html#command-opts">option</a>.</p></li>
<li><p><code>[sample_name]/[sample_name]_bedgraph_merged*.gz</code>: (When using the <code>--use_bme</code> <a href="command-opts.html#command-opts">option</a>) A <code>bedgraph</code> file containing positional and methylation information for each cytosine; a “coverage” file containing similar information. See <a href="https://rawgit.com/FelixKrueger/Bismark/master/Docs/Bismark_User_Guide.html#appendix-iii-bismark-methylation-extractor">Bismark’s documentation</a> for more detail.</p></li>
</ul>
<p><strong>Individual <em>bsseq</em> objects</strong> <code>BSobjects/objects/</code></p>
<ul>
<li><code>[chromosome]/[context]/*</code>: R objects containing either cytosines in “CpG” or “CpH”, for a single chromosome. While the <a href="outputs.html#main-outputs">final bsseq objects</a> are combined across all chromosomes for convenience, it may be more practical in some cases to work with these individual objects, which are significantly smaller in memory.</li>
</ul>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="inputs.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="configuration.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
