---
title: "Example Analysis using BiocMAP Output Objects"
author: 
  - name: Nicholas J. Eagles
    affiliation:
    - &libd Lieber Institute for Brain Development, Johns Hopkins Medical Campus
    email: nickeagles77@gmail.com
  - name: Leonardo Collado-Torres
    affiliation:
    - *libd
    - &ccb Center for Computational Biology, Johns Hopkins University
    email: lcolladotor@gmail.com
output:
  BiocStyle::pdf_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
---

```{r vignette_setup, echo=FALSE, message=FALSE, warning = FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()

## Bib setup
library("RefManageR")

## Write bibliography information
bib <- c(
    R = citation(),
    bsseq = citation("bsseq"),
    BiocStyle = citation("BiocStyle")[1],
    DelayedArray = citation("DelayedArray"),
    GGally = citation("GGally"),
    ggplot2 = citation("ggplot2"),
    HDF5Array = citation("HDF5Array"),
    here = citation("here"),
    knitr = citation("knitr")[1],
    RefManageR = citation("RefManageR")[1],
    rmarkdown = citation("rmarkdown")[1],
    sessioninfo = citation("sessioninfo")[1],
    tidyr = citation("tidyr")
)

options(width = 70)
```

BiocMAP is a computational pipeline for bisulfite-sequencing data. Starting from the raw FASTQ-based sequencer outputs, BiocMAP ultimately produces two `bsseq` R objects, containing methylation proportions and coverage information for all genomic cytosines in CpG and CpH methylation context. Various metrics, such as bisulfite-conversion rate, number of reads trimmed, or concordant alignment rate, are collected and become columns in the `colData` of the `bsseq` objects. The below analysis will demonstrate how to explore both methylation information and quality metrics in the BiocMAP output objects. Background understanding of the [bsseq](https://bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq.html) package, and its parent [SummarizedExperiment](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) are helpful but not required here.

The dataset used here includes 32 human postnatal dorsolateral prefrontal cortex samples up to 23 years of age. Samples were derived using NeuN-based fluorescence-activated nuclear sorting. The associated manuscript is [here](https://doi.org/10.1186/s13059-019-1805-1).

# Add Experiment Metadata to BiocMAP Outputs

The `bsseq` output objects from BiocMAP contain methylation and coverage info for our samples in the dataset. However, we're interested in exploring how this information relates back to sample metadata and phenotype information, present in an external file. Our first step will therefore be to load the BiocMAP output objects into memory, and manually attach the additional sample metadata to each object.

```{r, "load_packages", message = FALSE}
#  Load required R packages
library("bsseq")
library("HDF5Array")
library("ggplot2")
library("GGally")
library("tidyr")
library("here")
```

```{r, "define_inputs"}
#  Path to the sample metadata and BiocMAP outputs
meta_file <- here(
    "documentation", "example_analysis", "age_neun_pheno_data.csv"
)

#   Output bsseq objects from BiocMAP. A local directory is referenced because
#   the data is too large to host publicly
out_dir <- file.path(
    "/dcs04/lieber/lcolladotor/ageNeunSortedWGBS_LIBD001/ageNeunSortedWGBS",
    "processed-data/01-run_BiocMAP/pipeline_output/BSobjects/objects/combined"
)

#   CSV of DMRs already found for this dataset
dmr_list_path <- here(
    "documentation", "example_analysis", "age_neun_dmr_list.csv"
)
```

```{r setup, include = FALSE}
#   Only run the analysis when the BiocMAP files exist locally
knitr::opts_chunk$set(eval = dir.exists(out_dir))
```

```{r, "load_objects"}
#  Load the 'CpG'-context object
bs_cpg <- loadHDF5SummarizedExperiment(dir = out_dir, prefix = "CpG")

#  Load the 'CpH'-context object. Note: this requires quite a bit of memory
#  (~23GB) even though the assays are disk-backed!
bs_cph <- loadHDF5SummarizedExperiment(dir = out_dir, prefix = "CpH")

#  Read in experiment-specific metadata and ensure sample ID orders match
meta <- read.csv(meta_file)
meta <- meta[match(colnames(bs_cpg), meta$Data.ID), ]

#  Add this metadata to the Bioconductor objects
colData(bs_cpg) <- cbind(colData(bs_cpg), meta)
colData(bs_cph) <- cbind(colData(bs_cph), meta)

#  Keep a copy of the metadata as a data frame, for easy plotting
meta_df <- data.frame(colData(bs_cpg))
```

We'll briefly look at how the data is distributed by age and cell type shown in Table \@ref(tab:cellTypes) and Table \@ref(tab:ageBins), respectively.

```{r metaTables}
cell_type_df <- data.frame(table(meta$Cell.Type))
colnames(cell_type_df) <- c("Cell Type", "Num Samples")
knitr::kable(
    cell_type_df,
    caption = "Distribution of cell type across samples",
    label = "cellTypes"
)

age_bin_df <- data.frame(table(meta$Age.Bin))
colnames(age_bin_df) <- c("Age Bin", "Num Samples")
knitr::kable(
    age_bin_df,
    caption = "Distribution of age across samples",
    label = "ageBins"
)
```

# Exploratory Plots

## Bisulfite-Conversion Efficiency by Cell Type

This experiment used spike-ins of the lambda bacteriophage genome, which were quantified via BiocMAP to infer bisulfite-conversion rate. Successful bisulfite conversion is a pre-requisite for accurate methylation calls, so we'd like to see both that values (interpreted as percentages) are close to 100, and that values are not significantly different by sample (or by sample-related variables like cell type). We'll explore this visually in Figure \@ref(fig:Figure1).

```{r Figure1, warning = FALSE, out.width="100%", fig.align="center", fig.cap = "Bisulfite conversion rate by cell type across samples. Nearly all values are above 99\\% for both cell types, with one neuronal sample as low as just below 98.25\\%. Cell type does not appear to relate to bisulfite conversion rate, which is expected but reassuring."}
ggplot(data = meta_df, aes(x = Cell.Type, y = lambda_bs_conv_eff)) +
    geom_boxplot(outlier.shape = NA) +
    geom_point() +
    labs(x = "Cell Type", y = "Bisulfite Conversion Rate (%)") +
    theme_bw(base_size = 20)
```

## Relationship between Methylation Fractions across Cytosine Context, by Cell Type

Next, we'll explore if average methylation rate for each cytosine context correlates with that of other contexts across sample. For example, is a sample with highly methylated CpGs likely to have highly methylated CHGs? This is represented by the leftmost plot in Figure \@ref(fig:Figure2).

```{r Figure2, out.width="100%", fig.align="center", fig.cap = "Comparison of average methylation rate by cytosine context and cell type. CpG and CpH methylation of both types strongly correlate with each other, with an apparent linear relationship in all cases. Density plots reveal that methylation distributions have approximately the same shape but different means within each cytosine context between cell types, though variances can differ within each CpH context. Finally a dramatic difference in CpH methylation can be seen between neuronal and glial samples, which is consistent with the literature; CpH methylation is known to be more prominent in neurons than many other cell types."}
#   We'll make use of the 'ggpairs' function from the 'GGally' package, which is
#   well-suited for comparison of various metrics against each other, providing
#   density plots, comparison scatter plots, and correlation information.
ggpairs(
    data = meta_df,
    columns = c("perc_M_CpG", "perc_M_CHG", "perc_M_CHH"),
    xlab = "Methylation Rate (%)", ylab = "Methylation Rate (%)",
    columnLabels = c("CpG context", "CHG context", "CHH context"),
    mapping = aes(color = Cell.Type)
) +
    theme_bw(base_size = 15) +
    scale_color_brewer(palette = "Dark2") +
    scale_fill_brewer(palette = "Dark2")
```

## Proportion of Highly Methylated Cytosines Across Age by Cell Type and Context

Another interesting area of exploration would be to examine how methylation patterns change with age. Note that we are exploring methylation for different donors of different ages, as the data does not include different observations across time for a fixed donor. In particular, we'll look at the proportion of cytosines for a given trinucleotide context that are at least 10% methylated (across all observations for a fixed sample).

```{r "prepare_age_plot"}
#  The matrices in 'assays(bs_cpg)' and 'assays(bs_cph)' are stored on disk. To
#  speed up some below computations, we raise the per-block memory size
setAutoBlockSize(size = 1e9)

#  Get the proportion of cytosines in each object (context) that have > 10%
#  methylation
meta_df$high_meth_cpg <- DelayedArray::colMeans(assays(bs_cpg)$M > 0.1)
meta_df$high_meth_cph <- DelayedArray::colMeans(assays(bs_cph)$M > 0.1)

#  Convert the data frame to "long" format for use with ggplot
meta_df_long <- meta_df %>%
    pivot_longer(
        cols = starts_with("high_meth_"),
        names_to = "context",
        names_prefix = "high_meth_",
        values_to = "high_meth_c"
    )
```

```{r Figure3, out.width="100%", fig.align="center", fig.cap = "Proportion of highly methylated cytosines across age by cell type and context. While there isn't an apparent trend for CpG-context cytosines, CpH methylation show a clear increase with age, particularly for neurons."}
ggplot(
    data = meta_df_long,
    aes(x = Age, y = high_meth_c, color = Cell.Type, shape = context)
) +
    geom_point() +
    geom_smooth(method = "loess", aes(linetype = context), se = FALSE) +
    scale_shape_manual(values = c(19, 1), labels = c("CpG", "CpH")) +
    scale_color_brewer(palette = "Dark2") +
    labs(
        x = "Age", y = "Prop. Cytosines > 10% M", color = "Cell Type",
        shape = "Context"
    ) +
    guides(linetype = "none") +
    theme_bw(base_size = 20) +
    ylim(0, 1)
```

## Explore a DMR

The next natural area of analysis is computing differentially methylated regions (DMRs) between groups of interest. In this dataset, we might explore methylation differences between neurons and glia. Since the [corresponding manuscript](https://doi.org/10.1186/s13059-019-1805-1) already has done this, we'll read in the known DMRs from the supplementary table (additional file 2 [here](https://doi.org/10.1186/s13059-019-1805-1#additional-information)). Note that the `bsseq` package has [an excellent guide for computing DMRs](http://bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq_analysis.html#4_Finding_DMRs), a step that can be skipped in our case.

In this section, we'll read in the DMR list, "zoom out" around each DMR to form regions 100k bp wider (see `exploration_radius` below), and determine the region containing the most DMRs. We can then use `plotRegion` to visually check methylation levels at that region.

```{r prepare_dmr_plot}
exploration_radius <- 5e4

#   Read in CSV of DMRs and form a GenomicRanges object of unique regions
dmr_list <- read.csv(dmr_list_path)
dmr_gr <- GRanges(
    seqnames = dmr_list$Chromosome,
    ranges = IRanges(dmr_list$Start, dmr_list$End)
)
dmr_gr <- unique(dmr_gr)

#   Look around each DMR with a fixed radius
dmr_gr_wide <- resize(
    dmr_gr,
    width = width(dmr_gr) + 2 * exploration_radius, fix = "center"
)

#   Find the DMR with the most nearby DMRs
num_overlaps <- countOverlaps(dmr_gr, dmr_gr_wide)
dmr_gr_overlap <- dmr_gr[match(max(num_overlaps), num_overlaps)]

#   Color plots by cell type
pal = brewer.pal(3, "Dark2")[1:2]
p_data <- pData(bs_cpg)
p_data$col <- ifelse(bs_cpg$Cell.Type == "Glia", pal[1], pal[2])
pData(bs_cpg) <- p_data
```

```{r Figure4, out.width="100%", fig.align="center", fig.cap = "Genomic region containing differential methylation between neurons and glia. Orange methylation curves represent neuronal samples, whereas green curves represent glial samples. Windows highlighted in light red show DMRs determined in the manuscript between groups."}
plotRegion(
    BSseq = bs_cpg, region = dmr_gr_overlap,
    addRegions = dmr_gr, extend = exploration_radius
)
```

# Conclusion

The above analysis was intended to show the nature of the BiocMAP output objects and touch on the many possibilities for exploratory data analysis and downstream statistical processing. We recommend the [bsseq vignette](https://bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq_analysis.html) for additional guidance to complement the analysis shown here. More generally, we hope BiocMAP is useful in connecting researchers to the vast analysis possibilities made possible by the [hundreds of available Bioconductor packages](https://bioconductor.org/packages/release/BiocViews.html#___DNAMethylation).

# Reproducibility Info

Date the vignette was generated:

```{r reproduce1, echo=FALSE}
## Date the vignette was generated
Sys.time()
```

Wall-clock time spent generating the vignette:

```{r reproduce2, echo=FALSE}
## Processing time in seconds
totalTime <- diff(c(startTime, Sys.time()))
round(totalTime, digits = 3)
```

Memory-related information while generating the vignette:

```{r "memory"}
gc()
```

`R` session information:

```{r reproduce3, echo=FALSE}
## Session info
library("sessioninfo")
session_info()
```

# Bibliography

This vignette was made possible by the following packages and software:

* R `r Citep(bib[["R"]])`
* `r Biocpkg("bsseq")` = `r Citep(bib[["bsseq"]])`
* `r Biocpkg("BiocStyle")` = `r Citep(bib[["BiocStyle"]])`
* `r Biocpkg("DelayedArray")` = `r Citep(bib[["DelayedArray"]])`
* `r Biocpkg("GGally")` = `r Citep(bib[["GGally"]])`
* `r Biocpkg("ggplot2")` = `r Citep(bib[["ggplot2"]])`
* `r Biocpkg("HDF5Array")` = `r Citep(bib[["HDF5Array"]])`
* `r CRANpkg("here")` = `r Citep(bib[["here"]])`
* `r CRANpkg("knitr")` = `r Citep(bib[["knitr"]])`
* `r CRANpkg("RefManageR")` = `r Citep(bib[["RefManageR"]])`
* `r CRANpkg("rmarkdown")` = `r Citep(bib[["rmarkdown"]])`
* `r CRANpkg("sessioninfo")` = `r Citep(bib[["sessioninfo"]])`
* `r Biocpkg("tidyr")` = `r Citep(bib[["tidyr"]])`

```{r vignette_biblio, results = "asis", echo = FALSE, warning = FALSE, message = FALSE}
## Print bibliography
PrintBibliography(bib, .opts = list(hyperlink = "to.doc", style = "html"))
```
