---
title: "Example Analysis using BiocMAP Output Objects"
author: 
  - name: Nicholas J. Eagles
    affiliation:
    - &libd Lieber Institute for Brain Development, Johns Hopkins Medical Campus
    email: nickeagles77@gmail.com
  - name: Leonardo Collado-Torres
    affiliation:
    - *libd
    - &ccb Center for Computational Biology, Johns Hopkins University
    email: lcolladotor@gmail.com
output:
  BiocStyle::pdf_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
---

```{r vignette_setup, echo=FALSE, message=FALSE, warning = FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()

## Bib setup
library("RefManageR")

## Write bibliography information
bib <- c(
    R = citation(),
    bsseq = citation("bsseq"),
    BiocStyle = citation("BiocStyle")[1],
    DelayedArray = citation("DelayedArray"),
    GGally = citation("GGally"),
    ggplot2 = citation("ggplot2"),
    HDF5Array = citation("HDF5Array"),
    knitr = citation("knitr")[1],
    RefManageR = citation("RefManageR")[1],
    rmarkdown = citation("rmarkdown")[1],
    sessioninfo = citation("sessioninfo")[1],
    tidyr = citation("tidyr")
)

options(width = 70)
```

BiocMAP is a computational pipeline for bisulfite-sequencing data. Starting from the raw FASTQ-based sequencer outputs, BiocMAP ultimately produces two `bsseq` R objects, containing methylation proportions and coverage information for all genomic cytosines in CpG and CpH methylation context. Various metrics, such as bisulfite-conversion rate, number of reads trimmed, or concordant alignment rate, are collected and become columns in the `colData` of the `bsseq` objects. The below analysis will demonstrate how to explore both methylation information and quality metrics in the BiocMAP output objects. Background understanding of the [bsseq](https://bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq.html) package, and its parent [SummarizedExperiment](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) are helpful but not required here.

# Add Experiment Metadata to BiocMAP Outputs

The `bsseq` output objects from BiocMAP contain methylation and coverage info for our samples in the dataset. However, we're interested in exploring how this information relates back to sample metadata and phenotype information, present in an external file. Our first step will therefore be to load the BiocMAP output objects into memory, and manually attach the additional sample metadata to each object.

```{r, "load_packages", message = FALSE}
#  Load required R packages
library("bsseq")
library("HDF5Array")
library("ggplot2")
library("GGally")
library("tidyr")
```

```{r, "define_inputs"}
#  Path to the sample metadata and BiocMAP outputs. The outputs are too large
#  to host in this repository, so we reference local paths here
meta_file <- file.path(
    "/dcl02/lieber/ajaffe/FlowRNA_RNAseq/WGBS",
    "FlowRNA_WGBS_Sample_Information_with_Pheno_Info.csv"
)
out_dir <- file.path(
    "/dcs04/lieber/lcolladotor/flowRNA_LIBD001/flowRNA_WGBS/processed-data",
    "03_BiocMAP/BiocMAP_output"
)
```

```{r setup, include = FALSE}
#   Only run the analysis when the BiocMAP files exist locally
knitr::opts_chunk$set(eval = dir.exists(out_dir))
```

```{r, "load_objects"}
#  Load the 'CpG'-context object
bs_cpg <- loadHDF5SummarizedExperiment(
    file.path(out_dir, "BSobjects", "objects", "combined"),
    prefix = "CpG"
)

#  Load the 'CpH'-context object. Note: this requires quite a bit of memory
#  (~23GB) even though the assays are disk-backed!
bs_cph <- loadHDF5SummarizedExperiment(
    file.path(out_dir, "BSobjects", "objects", "combined"),
    prefix = "CpH"
)

#  Read in experiment-specific metadata and ensure sample ID orders match
meta <- read.csv(meta_file)
meta <- meta[match(colnames(bs_cpg), meta$LIBD.), ]

#  Add this metadata to the Bioconductor objects
colData(bs_cpg) <- cbind(colData(bs_cpg), meta)
colData(bs_cph) <- cbind(colData(bs_cph), meta)

#  Keep a copy of the metadata as a data frame, for easy plotting
meta_df <- data.frame(colData(bs_cpg))
```

# Exploratory Plots

## Bisulfite-Conversion Efficiency by Cell Population

This experiment used spike-ins of the lambda bacteriophage genome, which were quantified via BiocMAP to infer bisulfite-conversion rate. Successful bisulfite conversion is a pre-requisite for accurate methylation calls, so we'd like to see both that values (interpreted as percentages) are close to 100, and that values are not significantly different by sample (or by sample-related variables like cell population). We'll explore this visually below.

```{r Figure1, warning = FALSE, out.width="100%", fig.align="center", fig.cap = "Bisulfite conversion rate by cell population across samples. Values center at 95% for all cell populations, ranging just over 2% total. The 'TripleNeg' population might appear to have lower conversion rates, but it's unclear if the difference is significant."}
ggplot(meta_df, aes(x = Cell.Population, y = lambda_bs_conv_eff)) +
    geom_boxplot(outlier.shape = NA) +
    geom_point() +
    labs(x = "Cell Population", y = "Bisulfite Conversion Rate (%)") +
    theme_bw(base_size = 20)
```

## Relationship between Methylation Fractions across Cytosine Context, by Cell Population

Next, we'll explore if average methylation rate for each cytosine context correlates with that of other contexts across sample. For example, is a sample with highly methylated CpGs likely to have highly methylated CHGs? This is represented by the leftmost plot in middle row below.

```{r Figure2, out.width="100%", fig.align="center", fig.cap = "Comparison of average methylation rate by cytosine context and cell population. There is a correlation between average methylation rates of different cytosine contexts by sample. This is highly pronounced between CpH contexts (CHG vs. CHH). In each case, the relation appears roughly linear, though this is more questionable for the CpG vs. CHH context comparison. Generally, samples tend to cluster fairly well by cell population. Finally, for comparisons of CpG vs. CpH context, the strength of correlation between methylation rates varies significantly by cell type, with `MBPpos` showing only weak correlation at best."}
#   We'll make use of the 'ggpairs' function from the 'GGally' package, which is
#   well-suited for comparison of various metrics against each other, providing
#   density plots, comparison scatter plots, and correlation information.
ggpairs(
    meta_df,
    columns = c("perc_M_CpG", "perc_M_CHG", "perc_M_CHH"),
    xlab = "Methylation Rate (%)", ylab = "Methylation Rate (%)",
    columnLabels = c("CpG context", "CHG context", "CHH context"),
    mapping = aes(color = Cell.Population)
) +
    theme_bw(base_size = 20)
```

## Fraction of Covered Cytosines by Cell Population

Another useful piece of information is how well-covered the genome is with methylation information. Does coverage of cytosines vary by a sample's cell type?

```{r "coverage_fraction"}
#  The matrices in 'assays(bs_cpg)' and 'assays(bs_cph)' are stored on disk. To
#  speed up some below computations, we raise the per-block memory size
setAutoBlockSize(1e9)

#  Get the proportion of cytosines in each object (context) that have non-zero
#  coverage in at least one sample
meta_df$frac_covered_c_cpg <- DelayedArray::colMeans(assays(bs_cpg)$Cov > 3)
meta_df$frac_covered_c_cph <- DelayedArray::colMeans(assays(bs_cph)$Cov > 3)

#   Convert our data frame to "long" format, with a CpG-associated and
#   CpH-associated value for each coverage proportion
meta_df_long <- meta_df %>%
    pivot_longer(
        cols = starts_with("frac_covered_c"),
        names_to = "context",
        names_prefix = "frac_covered_c_",
        values_to = "frac_covered_c"
    )
```

```{r Figure3, out.width="100%", fig.align="center", fig.cap = "Proportion of covered cytosines by context and cell population. While around 80% of cytosines for a given sample have at least 3 observations in the experiment, there is wide variation between samples. At least 5 samples in each cytosine context have less than 50% sites 'well-covered', and CpH-context cytosines may have slightly less coverage generally than CpG-context cytosines."}
#   Plot fraction of covered cytosines by cell population
ggplot(
    meta_df_long,
    aes(x = Cell.Population, y = frac_covered_c, color = Cell.Population)
) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(position = "jitter") +
    facet_wrap(~context) +
    labs(
        title = "Proportion of Covered Cytosines by Cytosine Context and Cell Population",
        x = "Cell Population",
        y = "Proportion of Covered Cytosines"
    )  +
    theme_bw(base_size = 20)
```

## Distribution of Methylation Fractions across Cytosines by Cell Population

Grouping together all samples of a particular cell type, we'll explore the methylation-fraction distribution across cytosines, separately for both CpG and CpH contexts.

For the CpG cytosine context, we observe a bimodal distribution with peaks at fractions of 0 and 1. This suggests that within a particular sample, a CpG site is disproportionately likely to have consistent methylation pattern. For example, many CpG sites are such that all observations of the cytosine are methylated for a particular sample. Similarly, we don't see many sites where around half of the observed site are methylated for a given sample. It's also worth noting that the apparent bimodal form is likely not an artifact of low coverage-- i.e., only a small fraction of sites are covered just once or twice, a circumstance that would cause over-representation of the fractions of 0 or 1.

```{r "methylation_fractions"}
#  We'll use random sampling for computational efficiency. Set the seed for
#  reproducibility
set.seed(444)

#   Randomly subset to a particular number of cytosines, to both control memory
#   and speed up plotting
max_sites <- 1000

#   Look at CpG sites first
indices <- sample(nrow(bs_cpg), max_sites)
m_frac <- assays(bs_cpg)$M[indices, ] / assays(bs_cpg)$Cov[indices, ]

#   It's worth looking at the distribution of coverage by site, since in theory
#   this could be cause for the bimodality observed in the CpG portion of the
#   plot below
head(table(assays(bs_cpg)$Cov[indices, ]), n = 20)

#   Form a data frame for easy plotting: we'll collapse methylation data for all
#   samples into a single column, 'm_frac'. Here 'LIBD.' denotes sample ID
meth_df <- data.frame(
    "m_frac" = as.numeric(m_frac),
    "LIBD." = rep(colnames(m_frac), each = max_sites),
    "context" = "CpG"
)

#   Now look at CpH sites
indices <- sample(nrow(bs_cph), max_sites)
m_frac <- assays(bs_cph)$M[indices, ] / assays(bs_cph)$Cov[indices, ]

#   Again, we'll look at the distribution of coverage by site, to make sure it
#   isn't leading to noisy methylation results
head(table(assays(bs_cph)$Cov[indices, ]), n = 20)

#   Form a data frame for easy plotting: we'll collapse methylation data for all
#   samples into a single column, 'm_frac'. Here 'LIBD.' denotes sample ID. The
#   data frame will be appended to the existing CpG-associated one so we have
#   data for both cytosine contexts together.
meth_df <- rbind(
    meth_df,
    data.frame(
        "m_frac" = as.numeric(m_frac),
        "LIBD." = rep(colnames(m_frac), each = max_sites),
        "context" = "CpH"
    )
)

#   Label each observation (methylation fraction for a particular cytosine) with
#   the cell population of the associated sample
meth_df$Cell.Population <- meta_df$Cell.Population[
    match(meth_df$"LIBD.", meta_df$"LIBD.")
]
```

```{r Figure4, warning = FALSE, out.width="100%", fig.align="center", fig.cap = "Methylation proportion across cytosines by context and cell population. For the CpG cytosine context, we observe a bimodal distribution with peaks at fractions of 0 and 1. This suggests that within a particular sample, a CpG site is disproportionately likely to have consistent methylation pattern. CpH sites are far less methylated than CpG sites. This reflects the well-known fact about DNA methylation that most methylation occurs at CpGs, though in some cells (such as neurons), CpH methylation plays an important role as well." }
ggplot(meth_df, aes(x = Cell.Population, y = m_frac, color = Cell.Population)) +
    geom_violin() +
    facet_wrap(~context) +
    labs(
        title = "Methylation Proportion across Cytosines Cytosine Context and Cell Population",
        x = "Cell Population",
        y = "Methylation Proportion"
    )  +
    theme_bw(base_size = 20)
```

Next, we'll compute differentially methylated regions (DMRs) between cell populations. In particular, we'll compare MBP-positive samples to all other cell types, and plot methylation curves for CpG loci around the MBP genomic range.

```{r prepare_dmr_computation}
BiocParallel::register(BiocParallel::MulticoreParam(1))

cell_pop <- "MBPpos"

#   To avoid false positive DMRs, we'll subset to CpGs where at least 5 samples
#   in each group have at least two observations of the given CpG
num_cov_samples <- 5
num_cov_count <- 2

#   Subset object based on coverage requirements
bs_cov <- getCoverage(bs_cpg)
loci_to_keep <- which(
    rowSums(
        bs_cov[, bs_cpg$Cell.Population == cell_pop] >= num_cov_count
    ) >= num_cov_samples &
        rowSums(
            bs_cov[, bs_cpg$Cell.Population != cell_pop] >= num_cov_count
        ) >= num_cov_samples
)
length(loci_to_keep)

bs_cpg_sub <- bs_cpg[loci_to_keep, ]

#   Define the two cell-population-based groups used to compute t-statistics
group1 <- unique(bs_cpg_sub$"LIBD."[bs_cpg_sub$Cell.Population == cell_pop])
group2 <- unique(bs_cpg_sub$"LIBD."[bs_cpg_sub$Cell.Population != cell_pop])

#   Compute the t-stat and show the marginal distribution
cpg_t_stat <- BSmooth.tstat(
    bs_cpg_sub,
    group1 = group1, group2 = group2,
    estimate.var = "group2", local.correct = TRUE, verbose = TRUE
)
```

```{r Figure5, warning = FALSE, out.width="100%", fig.align="center", fig.cap = "Marginal distribution of t-statistics, before and after correction. [The bsseq vignette](http://bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq_analysis.html#3_Computing_t-statistics) recommends local correction to improve the symmetry of the marginal distribution, an effect that can be seen here."}
plot(cpg_t_stat)
```

```{r, "determine_dmrs"}
#   Grab the genomic range associated with this particular cell population
if (cell_pop == "SNAP25pos") {
    gene_range <- GRanges("chr20:10172395-10308258")
} else if (cell_pop == "MBPpos") {
    gene_range <- GRanges("chr18:76978827-77133708")
} else if (cell_pop == "GAD1pos") {
    gene_range <- GRanges("chr2:170813210-170861151")
} else {
    stop(
        paste0("No gene associated with this cell population '", cell_pop, "'.")
    )
}

t_stat_df <- getStats(cpg_t_stat)

#   Filters for the upcoming DMRs: minimum magnitude of t-stat and min number of
#   base pairs, respectively
thres <- sd(t_stat_df[, "tstat.corrected"]) * 2.5
n_bases <- 3
abs_mean_diff <- 0.1

#   Compute DMRs and apply above filters
dmrs_orig <- dmrFinder(cpg_t_stat, cutoff = c(-1 * thres, thres))
dmrs <- subset(dmrs_orig, n >= n_bases & abs(meanDiff) >= abs_mean_diff)

#   Color plots by cell population group
p_data <- pData(bs_cpg_sub)
p_data$col <- ifelse(bs_cpg_sub$Cell.Population == cell_pop, "red", "blue")
pData(bs_cpg_sub) <- p_data
```

```{r Figure6, out.width="100%", fig.align="center", fig.cap = "Methylation curves for MBP-positive samples (red) against other cell types (blue) around the MBP gene. To the eye, it appears that a fairly large region at the start of the MBP gene has differential methylation between groups; however, the only significant DMRs (highlighted in light red) are much narrrower bands dispersed throughout the gene (and somewhat less so outside of the gene)."}
#   Plot the region around the gene associated with this particular cell
#   population
plotRegion(
    bs_cpg_sub, gene_range,
    extend = width(gene_range) %/% 2, addRegions = dmrs
)
```

Date the vignette was generated:

```{r reproduce1, echo=FALSE}
## Date the vignette was generated
Sys.time()
```

Wall-clock time spent generating the vignette:

```{r reproduce2, echo=FALSE}
## Processing time in seconds
totalTime <- diff(c(startTime, Sys.time()))
round(totalTime, digits = 3)
```

`R` session information:

```{r reproduce3, echo=FALSE}
## Session info
library("sessioninfo")
session_info()
```

# Bibliography

This vignette was generated using `r Biocpkg("BiocStyle")` `r Citep(bib[["BiocStyle"]])`
with `r CRANpkg("knitr")` `r Citep(bib[["knitr"]])` and `r CRANpkg("rmarkdown")` `r Citep(bib[["rmarkdown"]])` running behind the scenes.

Citations made with `r CRANpkg("RefManageR")` `r Citep(bib[["RefManageR"]])`.

```{r vignette_biblio, results = "asis", echo = FALSE, warning = FALSE, message = FALSE}
## Print bibliography
PrintBibliography(bib, .opts = list(hyperlink = "to.doc", style = "html"))
```
