# All Command Options {#command-opts}

## BiocMAP Options

Each of the below parameters, unless otherwise noted, applies to both the first and second module provided in BiocMAP.

### Mandatory Parameters

* `--sample`        "single" or "paired": whether reads are single-end or paired-end
* `--reference`        "hg38", "hg19", or "mm10". The reference genome to be used for alignment and methylation extraction

### Optional Parameters

* `--all_alignments` (First module only) Include this flag to signal Arioc to also write outputs for discondant, rejected, and unmapped reads. Sam files for each outcome are kept as pipeline outputs. In either case, only concordant reads are used for later processing (methylation extraction and beyond)
* `--annotation`    The path to the directory containing reference-related files. Defaults to "./ref" (relative to the repository). If annotation files are not found here, the pipeline includes a step to build them.
* `--custom_anno [label]` Include this flag to indicate that the directory specified with `--annotation [dir]` includes user-provided annotation files to use instead of the default files. See the "Using custom annotation" section for more details.
* `--input`        If using the first module, the path to the directory containing the "samples.manifest" file, or, if using the second module, the path to the directory containing the "rules.txt" file. Defaults to "./test" (relative to the repository)
* `--output`  The path to the directory to store pipeline output files. Defaults to "./results" (relative to the repository)
* `--trim_mode`  (First module only) Determines the conditions under which trimming occurs:
                    "skip": do not perform trimming on samples
                    "adaptive": [default] perform trimming on samples that have failed the FastQC "Adapter content" metric
                    "force": perform trimming on all samples
* `--use_bme` (Second module only) Include this flag to perform methylation-extraction-related processes with Bismark utilities, rather than the default of MethylDackel
* `--with_lambda` (Second module only) Include this flag if all samples have spike-ins with the lambda bacteriophage genome. Pseudoalignment will then be performed to estimate bisulfite conversion efficiency

## Nextflow Options {#next-opts}

The nextflow command itself provides many additional options you may add to your "main" script. A few of the most commonly applicable ones are documented below. For a full list, type `[path to nextflow] run -h`- the full list does not appear to be documented at [nextflow's website](https://www.nextflow.io/).

* `-w [path]` Path to the directory where nextflow will place temporary files. This directory can fill up very quickly, especially for large experiments, and so it can be useful to set this to a scratch directory or filesystem with plenty of storage capacity.
* `-resume` Include this flag if pipeline execution halts with an error for any reason, and you wish to continue where you left off from last run. Otherwise, **by default, nextflow will restart execution from the beginning**.
* `-with-report [filename]` Include this to produce an html report with execution details (such as memory usage, completion details, and much more)
* `N [email address]` Sends email to the specified address to notify the user regarding pipeline completion. Note that nextflow relies on the `sendmail` tool for this functionality- therefore `sendmail` must be available for this option to work.
