# Annotation {#annotation}

BiocMAP can be run with `hg38`, `hg19`, or `mm10` references. The pipeline has a default and automated process for pulling and building annotation-related files, but the user can opt to provide their own annotation as an alternative. Both of these options are documented below. Example annotation files below are the ones used with default [configuration](#configuration) when `hg38` reference is selected.

- **A genome assembly fasta**: the reference genome to align reads to, like the file [here](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.primary_assembly.genome.fa.gz) (but unzipped).
- **Gene annotation gtf**: containing transcript data, like the file [here](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.annotation.gtf.gz) (but unzipped).
- **The lambda transcriptome**: for experiments utilizing spike-ins of the lambda bacteriophage genome, the transcriptome provided [here](ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/840/245/GCA_000840245.1_ViralProj14204/GCA_000840245.1_ViralProj14204_genomic.fna.gz) is used (but unzipped).

## Default Annotation

BiocMAP uses annotation files provided by [GENCODE](ftp://ftp.ebi.ac.uk/pub/databases/gencode/).

### Choosing a release ###

With genome annotation constantly being updated, the user may want to use a particular GENCODE release. The [configuration](#configuration) variables `gencode_version_human` and `gencode_version_mouse` control which GENCODE release is used for the human and mouse genomes, respectively.

```{groovy, eval=FALSE}
params {    
    //----------------------------------------------------
    //  Annotation-related settings
    //----------------------------------------------------
    
    gencode_version_human = "34"
    gencode_version_mouse = "M23"
    anno_build = "main" // main or primary (main is canonical seqs only)
```

### Choosing a "build" {#choosing-build}

Depending on the analysis you are doing, you may wish to only consider the reference chromosomes (for humans, the 25 sequences "chr1" through "chrM") for alignment and methylation extraction. BiocMAP provides the option to choose from two annotation "builds" for a given release and reference, called "main" and "primary" (following the naming convention from GENCODE databases).

* The "main" build consists of only the canonical "reference" sequences for each species
* The "primary" build consists of the canonical "reference" sequences and additional scaffolds, as a genome [primary assembly fasta from GENCODE](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_35/GRCh38.primary_assembly.genome.fa.gz) would contain.

See the variable `annotation_build` in your [configuration](#configuration) file for making this selection for your pipeline run.

## Custom Annotation

You may wish to provide a genome FASTA (the reference genome to align reads to), such as the file [here](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.primary_assembly.genome.fa.gz), in place of the automatically managed GENCODE files described in the above section.

You must also add the `--custom_anno [label]` argument to your [execution scripts](#quick-config), to specify you are using custom annotation files. The "label" is a string you want to include in filenames generated from the annotation files you provided. This is intended to allow the use of potentially many different custom annotations, assigned a unique and informative name you choose each time. This can be anything except an empty string (which internally signifies not to use custom annotation).
