# Outputs {#outputs}

## Main Outputs

### First Module

[TODO]

### Second Module

[LIKELY WANT A FIGURE DEPICTING OUTPUT OBJECTS]

#### R Objects

[TODO]

#### Metrics

[TODO]

## Intermediary Outputs

[INSERT PIPELINE NAME HERE] generates a number of files along the process before producing the main outputs of interest in each module. Each of these "intermediary" files is described below.

### First Module

[TODO- needs editing and additional content]

**FastQC Outputs** `fastQC/`

- `[trim_status]/[file_name]/*`: Outputs from [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). Here `[trim_status]` indicates when FastQC was performed: `Untrimmed` is before trimming, and `Trimmed` is after. `[file_name]` contains the sample name, and if applicable, the mate number.

**Alignment BAMs and Summaries** `alignment/`

- `[sample_name].bam`: The main alignment output from [HISAT2](http://daehwankimlab.github.io/hisat2/) or optionally [STAR](https://github.com/alexdobin/STAR) in BAM format. In either case, unmapped reads are not included (different from the default behavior for HISAT2!).
- `[sample_name]_align_summary.txt`: The text-based alignment summary from [Hisat2](http://daehwankimlab.github.io/hisat2/), if applicable. Note that metrics from these files are aggregated for the experiment, and so users likely will not need to check or process the original files manually.
- `[sample_name]_STAR_alignment.log`: Statistics from STAR alignment (if applicable) for a single sample, renamed from `Log.final.out`. Note that metrics from these files are aggregated for the experiment, and so users likely will not need to check or process the original files manually.
- `bam_sort/[sample_name]_sorted.bam` and `bam_sort/[sample_name]_sorted.bam.bai`: Coordinate-sorted alignments and their corresponding indices.
- `[sample_name]_unmapped_mate*.fastq`: If using STAR with paired-end reads and the `--unalign` [option](#command-opts), these two files (mates 1 and 2) are produced, including unmapped reads (this includes "discordant" reads).
- `[sample_name]_discordant.fastq`: If using HISAT2 (this is the default) and the `--unalign` [option](#command-opts), this file is produced and includes discordant mappings.

**Trimmed FASTQ Files** `trimming/`

- `[sample_name]_trimmed*.fastq`: Trimmed FASTQ files, if applicable, from [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)

### Second Module

**Preprocessing Logs** `preprocessing/`

- `preprocess_input_second_half.log`: Information about how the input `rules.txt` file was parsed to internally handle input files correctly downstream.

**Lambda-based Bisulfite-Conversion Estimatation** `lambda/`

- `[sample_name]_lambda_pseudo.log`: (For experiments with lambda spike-ins, using the `--with_lambda` [option](#command-opts)) Logs containing a percentage estimate for the bisulfite-conversion efficiency for each sample, based on pseudo-aligning input FASTQs to both the lambda bacteriophage transcriptome and an "*in silico* bisulfite-converted copy" of the transcriptome, and comparing alignment rates.

**Bismark Methylation Extractor Outputs** `BME/`

- `[sample_name]/*`: (When using the `--use_bme` [option](#command-opts)) Files output from running `bismark_methylation_extractor`, including compressed text-based methylation information, a splitting report, and an "M-bias" report.

**Cytosine Reports and Bedgraphs** `Reports/`

- `[sample_name]/[sample_name].*.CX_report.txt`: "Cytosine reports" containing methylation counts and context for every cytosine in the genome, split by chromosome. By default, these are generated with `MethylDackel`, or otherwise with `coverage2cytosine` from the `Bismark` software suite when using the `--use_bme` [option](#command-opts).

- `[sample_name]/[sample_name]_bedgraph_merged*.gz`: (When using the `--use_bme` [option](#command-opts)) A `bedgraph` file containing positional and methylation information for each cytosine; a "coverage" file containing similar information. See [Bismark's documentation](https://rawgit.com/FelixKrueger/Bismark/master/Docs/Bismark_User_Guide.html#appendix-iii-bismark-methylation-extractor) for more detail.
