# Outputs {#outputs}

## Main Outputs

### First Module

[TODO]

### Second Module

[LIKELY WANT A FIGURE DEPICTING OUTPUT OBJECTS]

#### R Objects

The major outputs from the second module are R objects from the [*bsseq* Bioconductor package](https://www.bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq.html), which contain methylation proportion and coverage information at all cytosine loci in the human genome. bsseq extends the `SummarizedExperiment` class, which provides a general and popular format for storing genomics data. Two `bsseq` objects are produced, with one object containing cytosine sites in CpG context, and the other containing the remaining CpH loci.

We "strand collapse" CpG loci, which involves combining methylation data from both genomic strands (and thus discarding strand-specific information). The object is "smoothed" with the [BSmooth algorithm](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r83), a process for inferring stable regional estimates of methylation levels. Loci are ordered by genomic position.

We retain strand-specific information for CpH loci. These loci are also ordered by genomic position.

Because all cytosines in the genome are included in the output objects, the data may occupy tens or even hundreds of gigabytes in memory (RAM) if loaded in a typical fashion. To enable working with the objects in a reasonable amount of memory, the assays (in this case methylation fraction and coverage counts) are HDF5-backed using the [HDF5Array R package](https://bioconductor.org/packages/release/bioc/html/HDF5Array.html), based on the [DelayedArray](https://bioconductor.org/packages/release/bioc/html/DelayedArray.html) framework. Essentially, this involves storing the assays in a .h5 file, a format designed to enable working with on-disk data as if it were loaded in RAM.

#### Metrics

[TODO]

## Intermediary Outputs

[INSERT PIPELINE NAME HERE] generates a number of files along the process before producing the main outputs of interest in each module. Each of these "intermediary" files is described below.

### First Module

**Preprocessing Logs** `preprocessing/`

- `preprocess_input_first_half.log`: Information about how the input `samples.manifest` file was parsed to internally handle input files correctly downstream.

**FastQC Outputs** `fastQC/`

- `[trim_status]/[file_name]/*`: Outputs from [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). Here `[trim_status]` indicates when FastQC was performed: `Untrimmed` is before trimming, and `Trimmed` is after. `[file_name]` contains the sample name, and if applicable, the mate number.

**Trimmed FASTQ Files** `trimming/`

- `[sample_name]_trimmed*.fastq`: Trimmed FASTQ files, if applicable, from [Trim Galore](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/).

**Alignment-related Files** `Arioc/`

- `sams/[sample_name].c.sam`: Concordant alignments output directly from Arioc, in SAM format.
- `configs/[sample_name]_align.cfg`: The exact config file used by Arioc to align reads to the reference genome.
- `configs/[sample_name]_encode.cfg`: The exact config file used by Arioc to encode reads prior to alignment to the reference genome.
- `logs/*_[sample_name].log`: Logs from encoding and aligning samples with Arioc.

### Second Module

**Preprocessing Logs** `preprocessing/`

- `preprocess_input_second_half.log`: Information about how the input `rules.txt` file was parsed to internally handle input files correctly downstream.

**Lambda-based Bisulfite-Conversion Estimatation** `lambda/`

- `[sample_name]_lambda_pseudo.log`: (For experiments with lambda spike-ins, using the `--with_lambda` [option](#command-opts)) Logs containing a percentage estimate for the bisulfite-conversion efficiency for each sample, based on pseudo-aligning input FASTQs to both the lambda bacteriophage transcriptome and an "*in silico* bisulfite-converted copy" of the transcriptome, and comparing alignment rates.

**Bismark Methylation Extractor Outputs** `BME/`

- `[sample_name]/*`: (When using the `--use_bme` [option](#command-opts)) Files output from running `bismark_methylation_extractor`, including compressed text-based methylation information, a splitting report, and an "M-bias" report.

**Cytosine Reports and Bedgraphs** `Reports/`

- `[sample_name]/[sample_name].*.CX_report.txt`: "Cytosine reports" containing methylation counts and context for every cytosine in the genome, split by chromosome. By default, these are generated with `MethylDackel`, or otherwise with `coverage2cytosine` from the `Bismark` software suite when using the `--use_bme` [option](#command-opts).

- `[sample_name]/[sample_name]_bedgraph_merged*.gz`: (When using the `--use_bme` [option](#command-opts)) A `bedgraph` file containing positional and methylation information for each cytosine; a "coverage" file containing similar information. See [Bismark's documentation](https://rawgit.com/FelixKrueger/Bismark/master/Docs/Bismark_User_Guide.html#appendix-iii-bismark-methylation-extractor) for more detail.
