---
title: "BiocMAP"
author:
  - name: Nicholas J. Eagles
    affiliation:
    - &libd Lieber Institute for Brain Development, Johns Hopkins Medical Campus
  - name: Leonardo Collado-Torres
    affiliation:
    - *libd
    - &ccb Center for Computational Biology, Johns Hopkins University
    email: lcolladotor@gmail.com
site: bookdown::bookdown_site
documentclass: book
output:
  bookdown::gitbook: default
  #bookdown::pdf_book: default
---

# Overview {-}

BiocMAP is a pair of [Nextflow](https://www.nextflow.io/)-based pipelines for processing raw whole genome bisulfite sequencing data into [Bioconductor](https://bioconductor.org/)-friendly [bsseq](https://www.bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq.html) objects in R.

The first BiocMAP module performs speedy alignment to a reference genome by [Arioc](https://github.com/RWilton/Arioc), and requires GPU resources. Methylation extraction and remaining steps are performed in the second module, optionally on a different computing system where GPUs need not be available.

![](images/workflow.png)
*Diagram representing the "conceptual" workflow traversed by BiocMAP. Here some nextflow [`processes`](https://www.nextflow.io/docs/latest/process.html) are grouped together for simplicity; the exact processes traversed are enumerated [here](#pipeline-overview). The red box indicates the FASTQ files are inputs to the pipeline; green coloring denotes major output files from the pipeline; the remaining boxes represent computational steps. Yellow-colored steps are optional or not always performed; for example, lambda pseudoalignment is an optional step intended for experiments with spike-ins of the lambda bacteriophage. Finally, blue-colored steps are ordinary processes which occur on every pipeline execution.*

## Cite `BiocMAP` {-}

We hope `BiocMAP` will be a useful tool for your research. Please use the following bibtex information to cite this software. Thank you!

```
@article {Eagles2022.04.20.488947,
	author = {Eagles, Nicholas J and Wilton, Richard and Jaffe, Andrew E. and Collado-Torres, Leonardo},
	title = {BiocMAP: A Bioconductor-friendly, GPU-Accelerated Pipeline for Bisulfite-Sequencing Data},
	year = {2022},
	doi = {10.1101/2022.04.20.488947},
	publisher = {Cold Spring Harbor Laboratory},
	URL = {https://doi.org/10.1101/2022.04.20.488947},
	journal = {bioRxiv}
}
```

This is a project by the [R/Bioconductor-powered Team Data Science](https://lcolladotor.github.io/bioc_team_ds) at the [Lieber Institute for Brain Development](https://www.libd.org/).


|     |     |
|-----|-----|
| <a href="https://lcolladotor.github.io/bioc_team_ds"><img src="http://lcolladotor.github.io/img/Leo_transparente.png" width="250px"></a> | <a href="https://www.libd.org/"><img src="http://lcolladotor.github.io/img/LIBD_logo.jpg" width="250px"></a> |

## R session information {-}

Details on the R version used for making this book. The source code is available at [`LieberInstitute/BiocMAP`](https://github.com/LieberInstitute/BiocMAP).

```{r session_info_chunk, eval = FALSE}
## Load the package at the top of your script
library("sessioninfo")

## Reproducibility information
print('Reproducibility information:')
Sys.time()
proc.time()
options(width = 120)
session_info()
```

```{r session_info, echo = FALSE}
<<session_info_chunk>>
```

This book was last updated on `r Sys.time()`.
